File: kotlin/src/main/kotlin/chromahash/MuLaw.kt
package chromahash

import kotlin.math.abs
import kotlin.math.ln
import kotlin.math.pow
import kotlin.math.sign

/** mu-law compress: value in [-1, 1] -> compressed in [-1, 1]. */
internal fun muCompress(value: Double): Double {
    val v = value.coerceIn(-1.0, 1.0)
    return sign(v) * ln(1.0 + MU * abs(v)) / ln(1.0 + MU)
}

/** mu-law expand: compressed in [-1, 1] -> value in [-1, 1]. */
internal fun muExpand(compressed: Double): Double = sign(compressed) * ((1.0 + MU).pow(abs(compressed)) - 1.0) / MU

/** Quantize a value in [-1, 1] using mu-law to an integer index. */
internal fun muLawQuantize(
    value: Double,
    bits: Int,
): Int {
    val compressed = muCompress(value)
    val maxVal = (1 shl bits) - 1
    val index = roundHalfAwayFromZero((compressed + 1.0) / 2.0 * maxVal.toDouble())
    return index.toLong().coerceIn(0L, maxVal.toLong()).toInt()
}

/** Dequantize an integer index back to a value in [-1, 1] using mu-law. */
internal fun muLawDequantize(
    index: Int,
    bits: Int,
): Double {
    val maxVal = (1 shl bits) - 1
    val compressed = index.toDouble() / maxVal.toDouble() * 2.0 - 1.0
    return muExpand(compressed)
}

File: csharp/src/Chromahash/MuLaw.cs
namespace ChromaHash;

using static Constants;
using static MathUtils;

internal static class MuLaw
{
    /// <summary>µ-law compress: value in [-1, 1] → compressed in [-1, 1].</summary>
    public static double MuCompress(double value)
    {
        double v = Math.Clamp(value, -1.0, 1.0);
        return Math.Sign(v) * Math.Log(1.0 + Mu * Math.Abs(v)) / Math.Log(1.0 + Mu);
    }

    /// <summary>µ-law expand: compressed in [-1, 1] → value in [-1, 1].</summary>
    public static double MuExpand(double compressed)
    {
        return Math.Sign(compressed) * (Math.Pow(1.0 + Mu, Math.Abs(compressed)) - 1.0) / Mu;
    }

    /// <summary>Quantize a value in [-1, 1] using µ-law to an integer index. Per spec §12.7 muLawQuantize.</summary>
    public static uint MuLawQuantize(double value, int bits)
    {
        double compressed = MuCompress(value);
        uint maxVal = (1u << bits) - 1;
        double index = RoundHalfAwayFromZero((compressed + 1.0) / 2.0 * maxVal);
        return (uint)Math.Clamp((long)index, 0L, (long)maxVal);
    }

    /// <summary>Dequantize an integer index back to a value in [-1, 1] using µ-law. Per spec §12.7 muLawDequantize.</summary>
    public static double MuLawDequantize(uint index, int bits)
    {
        uint maxVal = (1u << bits) - 1;
        double compressed = (double)index / maxVal * 2.0 - 1.0;
        return MuExpand(compressed);
    }
}
